File: src/module/classify/05/33.py
class Solution:
    # 153. 寻找旋转排序数组中的最小值
    def findMin(self, nums: List[int]) -> int:
        left, right = -1, len(nums) - 1  # 开区间 (-1, n-1)
        while left + 1 < right:  # 开区间不为空
            mid = (left + right) // 2
            if nums[mid] < nums[-1]:  # 蓝色
                right = mid
            else:  # 红色
                left = mid
        return right

    # 有序数组中找 target 的下标
    def lower_bound(self, nums: List[int], left: int, right: int, target: int) -> int:
        while left + 1 < right:  # 开区间不为空
            mid = (left + right) // 2
            # 循环不变量：
            # nums[left] < target
            # nums[right] >= target
            if nums[mid] < target:
                left = mid  # 范围缩小到 (mid, right)
            else:
                right = mid  # 范围缩小到 (left, mid)
        return right if nums[right] == target else -1

    def search(self, nums: List[int], target: int) -> int:
        i = self.findMin(nums)
        if target > nums[-1]:  # target 在第一段
            return self.lower_bound(nums, -1, i, target)  # 开区间 (-1, i)
        # target 在第二段
        return self.lower_bound(nums, i - 1, len(nums), target)  # 开区间 (i-1, n)

File: src/module/classify/05/33.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
  let begin = 0,
    end = nums.length - 1;
  while (begin <= end) {
    // 下面这样写是考虑大数情况下避免溢出
    let mid = begin + ((end - begin) >> 1);
    if (nums[mid] == target) {
      return mid;
    } else {
      // 细节点在于里面的判断全部都是<= 或>= 而不是<或>
      // 如果左半部分是有序的的
      if (nums[begin] <= nums[mid]) {
        //target在[ nums[begin],nums[mid] ]中，那么就在这段有序区间查找
        if (nums[begin] <= target && target <= nums[mid]) {
          end = mid - 1;
        } else {
          //否则去反方向查找
          begin = mid + 1;
        }
      } else {
        // 右半部分是有序的
        //target在[ nums[mid],nums[end] ]中，那么就在这段有序区间查找
        if (nums[mid] <= target && target <= nums[end]) {
          begin = mid + 1;
        } else {
          end = mid - 1;
        }
      }
    }
  }
  return -1;
};